common/include/villas/kernel/devices/device_connection.hpp

```diff
@@ -0,0 +1,79 @@
+/* Vfio connection to a device.
+ *
+ * Author: Pascal Bauer <pascal.bauer@rwth-aachen.de>
+ *
+ * SPDX-FileCopyrightText: 2024-25 Pascal Bauer <pascal.bauer@rwth-aachen.de>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <memory>
+
+#include <villas/kernel/devices/linux_driver.hpp>
+#include <villas/kernel/vfio_device.hpp>
+
+namespace villas {
+namespace kernel {
+namespace devices {
+
+class DeviceConnection {
+public:
+  Logger logger;
+  const std::shared_ptr<kernel::vfio::Device> vfio_device;
+
+private:
+  DeviceConnection(std::shared_ptr<kernel::vfio::Device> vfio_device)
+      : logger(villas::Log::get("DeviceConnection")),
+        vfio_device(vfio_device) {};
+
+public:
+  static DeviceConnection
+  from(const villas::kernel::devices::Device &device,
+       std::shared_ptr<kernel::vfio::Container> vfio_container) {
+    auto logger = villas::Log::get("Builder: DeviceConnection");
+
+    // Bind to driver
+    LinuxDriver driver(
+        std::filesystem::path("/sys/bus/platform/drivers/vfio-platform"));
+    driver.attach(device);
+
+    // Attach group to container
+    const int iommu_group = device.iommu_group().value();
+    auto vfio_group = vfio_container->getOrAttachGroup(iommu_group);
+    logger->debug("Device: {}, Iommu: {}", device.name(), iommu_group);
+
+    // Open Vfio Device
+    auto vfio_device = std::make_shared<kernel::vfio::Device>(
+        device.name(), vfio_group->getFileDescriptor());
+
+    // Attach device to group
+    vfio_group->attachDevice(vfio_device);
+
+    return DeviceConnection(vfio_device);
+  }
+
+  void add_to_memorygraph() const {
```

Comment:
Please note that the LLVM code-style is mandating camelCase for function names:

https://releases.llvm.org/10.0.0/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly